WalletWasabi.Fluent/Models/Wallets/AddressesModel.cs

```diff
@@ -1,47 +1,69 @@
-using System.Reactive.Disposables;
+using System.Reactive.Linq;
+using System.Reactive.Subjects;
 using DynamicData;
 using WalletWasabi.Blockchain.Keys;
-using WalletWasabi.Fluent.Extensions;
+using WalletWasabi.Blockchain.TransactionProcessing;
+using WalletWasabi.Wallets;
 
 namespace WalletWasabi.Fluent.Models.Wallets;
 
 [AutoInterface]
-public partial class AddressesModel : IDisposable
+public partial class AddressesModel
 {
-	private readonly CompositeDisposable _disposable = new();
-	private readonly KeyManager _keyManager;
+	private readonly ISubject<HdPubKey> _newAddressGenerated = new Subject<HdPubKey>();
+	private readonly Wallet _wallet;
+	private readonly SourceList<HdPubKey> _source;
 
-	public AddressesModel(IObservable<Unit> addressesUpdated, KeyManager keyManager)
+	public AddressesModel(Wallet wallet)
 	{
-		_keyManager = keyManager;
+		_wallet = wallet;
+		_source = new SourceList<HdPubKey>();
+		_source.AddRange(GetUnusedKeys());
 
-		Cache =
-			addressesUpdated.Fetch(GetAddresses, address => address.Text)
-								.DisposeWith(_disposable);
+		Observable.FromEventPattern<ProcessedResult>(
+				h => wallet.WalletRelevantTransactionProcessed += h,
+				h => wallet.WalletRelevantTransactionProcessed -= h)
+			.Do(_ => UpdateUnusedKeys())
+			.Subscribe();
 
-		UnusedAddressesCache =
-			Cache.Connect()
-				 .AutoRefresh(x => x.IsUsed)
-				 .Filter(x => !x.IsUsed)
-				 .AsObservableCache()
-				 .DisposeWith(_disposable);
+		_newAddressGenerated
+			.Do(address => _source.Add(address))
+			.Subscribe();
 
-		HasUnusedAddresses = UnusedAddressesCache.NotEmpty();
+		_source.Connect()
+			.Transform(key => (IAddress) new Address(this, _wallet.KeyManager, key))
+			.Bind(out var unusedAddresses)
+			.Subscribe();
+
+		Unused = unusedAddresses;
 	}
 
-	public IObservableCache<IAddress, string> Cache { get; }
+	private IEnumerable<HdPubKey> GetUnusedKeys() => _wallet.KeyManager.GetKeys(x => x is { IsInternal: false, KeyState: KeyState.Clean, Labels.Count: > 0 });
+
+	public IAddress NextReceiveAddress(IEnumerable<string> destinationLabels)
+	{
+		var pubKey = _wallet.GetNextReceiveAddress(destinationLabels);
+		var nextReceiveAddress = new Address(this, _wallet.KeyManager, pubKey);
+		_newAddressGenerated.OnNext(pubKey);
 
-	public IObservableCache<IAddress, string> UnusedAddressesCache { get; }
+		return nextReceiveAddress;
+	}
 
-	public IObservable<bool> HasUnusedAddresses { get; }
+	public ReadOnlyObservableCollection<IAddress> Unused { get; }
 
-	public void Dispose() => _disposable.Dispose();
+	public void Hide(Address address)
+	{
+		_wallet.KeyManager.SetKeyState(KeyState.Locked, address.HdPubKey);
+		_wallet.KeyManager.ToFile();
+		_source.Remove(address.HdPubKey);
+	}
 
-	private IEnumerable<IAddress> GetAddresses() => _keyManager
-		.GetKeys()
-		.Reverse()
-		.Select(x => new Address(_keyManager, x));
+	private void UpdateUnusedKeys()
+	{
+		var hdPubKeys = GetUnusedKeys().ToList();
+		_source.EditDiff(hdPubKeys);
```

Comment:
That was my original code, but I ended up with this version to ensure that items are up to date. I had some problem getting updated items, so I implemented this as a fallback and forgot to update.

So, correct me if I'm wrong: HdPubKeys obtained by GetKeys are mutated by <someone> to keep in sync before the WalletRelevantTransaction event is raised.

Well, I ask you this because I tried to locate the point in the code where the KeyState is mutated, but it was so time consuming that I gave up.